README.md
# npme-auth-oauth2

[![Build Status](https://travis-ci.org/npm/npme-auth-oauth2.png)](https://travis-ci.org/npm/npme-auth-oauth2)
[![Coverage Status](https://coveralls.io/repos/npm/npme-auth-oauth2/badge.svg?branch=master)](https://coveralls.io/r/npm/npme-auth-oauth2?branch=master)

auth strategy for OAuth2 SSO.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.2.0"></a>
# [1.2.0](https://github.com/npm/npme-auth-oauth2/compare/v1.1.0...v1.2.0) (2016-12-19)


### Features

* indicate that upstream caching should not be allowed ([#2](https://github.com/npm/npme-auth-oauth2/issues/2)) ([39c6bc3](https://github.com/npm/npme-auth-oauth2/commit/39c6bc3))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/npm/npme-auth-oauth2/compare/v1.0.0...v1.1.0) (2016-10-14)


### Features

* lock an auth token for a few minutes ([#1](https://github.com/npm/npme-auth-oauth2/issues/1)) ([45fb22a](https://github.com/npm/npme-auth-oauth2/commit/45fb22a))



<a name="1.0.0"></a>
# 1.0.0 (2016-08-04)


### Bug Fixes

* oauth library no longer supports node 0.10 ([7188369](https://github.com/npm/npme-auth-oauth2/commit/7188369))


### Features

* built out an OAuth 2.0 auth strategy ([2aa722a](https://github.com/npm/npme-auth-oauth2/commit/2aa722a))


test/session.js
var client = require('redis').createClient()
var Session = require('../lib/session')
var tap = require('tap')

var sessionLookupKey = 'user-abc123'
var user = {name: 'bcoe', email: 'ben@example.com'}

process.env.OAUTH2_AUTHORIZATION_PATH = 'https://auth.example.com'

function deleteTokens (t) {
  client.del(sessionLookupKey, function () {
    t.done()
  })
}

tap.test('before', deleteTokens)

tap.test('it should return an error containing a login URL if no session exists for a user', function (t) {
  var session = new Session()

  session.get(sessionLookupKey, function (err, user) {
    session.end()
    t.equal(err.statusCode, 401)
    t.ok(err.message.indexOf('https://auth.example.com') !== -1)
    t.done()
  })
})

tap.test('before', deleteTokens)

tap.test('it should allow a session to be set for a user', function (t) {
  var session = new Session()

  session.set(sessionLookupKey, user, function (err) {
    t.false(err)
    client.get(sessionLookupKey, function (err, res) {
      session.end()

      t.equal(err, null)
      t.ok(res.indexOf('ben@example.com') !== -1)
      t.done()
    })
  })
})

tap.test('it should allow a session to be fetched for a user', function (t) {
  var session = new Session()

  session.get(sessionLookupKey, function (err, user) {
    session.end()
    t.false(err)
    t.equal(user.email, 'ben@example.com')
    t.done()
  })
})

tap.test('before', deleteTokens)

tap.test('after', function (t) {
  client.end(true)
  t.done()
})


test/authorizer.js
var Authorizer = require('../lib/authorizer')
var Session = require('../lib/session')
var tap = require('tap')
var nock = require('nock')

var userComplete = {name: 'bcoe', email: 'ben@example.com', accessToken: 'abc123'}
var userNotComplete = {name: 'bcoe', email: 'ben@example.com'}

var session = new Session()

process.env.OAUTH2_AUTHORIZATION_PATH = 'https://auth.example.com'
process.env.OAUTH2_PROFILE = 'https://api.github.com/user'

tap.test('it responds with session object if SSO dance is complete', function (t) {
  var authorizer = new Authorizer()
  var profile = nock('https://api.github.com')
    .get('/user')
    .reply(200)

  session.unlock('ben@example.com-abc123')

  session.set('ben@example.com-abc123', userComplete, function (err) {
    t.equal(err, null)
    authorizer.authorize({
      headers: {
        authorization: 'Bearer ben@example.com-abc123'
      }
    }, function (err, user) {
      authorizer.end()
      session.unlock('ben@example.com-abc123')
      session.delete('ben@example.com-abc123')

      profile.done()
      t.equal(err, null)
      t.equal(user.email, 'ben@example.com')
      t.end()
    })
  })
})

tap.test('does not check github if request is within timeout window', function (t) {
  var authorizer = new Authorizer()

  session.lock('ben@example.com-abc123')
  session.set('ben@example.com-abc123', userComplete, function (err) {
    t.equal(err, null)
    authorizer.authorize({
      headers: {
        authorization: 'Bearer ben@example.com-abc123'
      }
    }, function (err, user) {
      authorizer.end()
      session.delete('ben@example.com-abc123')
      session.unlock('ben@example.com-abc123')

      t.equal(err, null)
      t.equal(user.email, 'ben@example.com')
      t.end()
    })
  })
})

tap.test('it returns error with login url if access token is no longer valid', function (t) {
  var authorizer = new Authorizer()
  var profile = nock('https://api.github.com')
    .get('/user')
    .reply(401)

  session.set('ben@example.com-abc123', userComplete, function (err) {
    t.equal(err, null)
    authorizer.authorize({
      headers: {
        authorization: 'Bearer ben@example.com-abc123'
      }
    }, function (err, user) {
      authorizer.end()
      session.delete('ben@example.com-abc123')
      session.unlock('ben@example.com-abc123')

      profile.done()
      t.ok(err.message.indexOf('visit https://auth.example.com') !== -1)
      t.end()
    })
  })
})

tap.test('it returns error with login url if SSO dance is not complete', function (t) {
  var authorizer = new Authorizer()
  session.set('ben@example.com-abc123', userNotComplete, function (err) {
    t.equal(err, null)
    authorizer.authorize({
      headers: {
        authorization: 'Bearer ben@example.com-abc123'
      }
    }, function (err, user) {
      authorizer.end()
      session.delete('ben@example.com-abc123')
      session.unlock('ben@example.com-abc123')

      t.ok(err.message.indexOf('visit https://auth.example.com') !== -1)
      t.end()
    })
  })
})

tap.test('it indicates that upstream caching is not allowed with OAuth-verified response', t => {
  var authorizer = new Authorizer()
  var profile = nock('https://api.github.com')
    .get('/user')
    .reply(200)

  session.unlock('ben@example.com-abc123')

  session.set('ben@example.com-abc123', userComplete, function (err) {
    t.equal(err, null)
    authorizer.authorize({
      headers: {
        authorization: 'Bearer ben@example.com-abc123'
      }
    }, function (err, user) {
      authorizer.end()
      session.unlock('ben@example.com-abc123')
      session.delete('ben@example.com-abc123')

      profile.done()
      t.equal(err, null)
      t.equal(user.email, 'ben@example.com')
      t.equal(user.cacheAllowed, false)
      t.end()
    })
  })
})

tap.test('it indicates that upstream caching is not allowed with cached response', t => {
  var authorizer = new Authorizer()

  session.lock('ben@example.com-abc123')
  session.set('ben@example.com-abc123', userComplete, function (err) {
    t.equal(err, null)
    authorizer.authorize({
      headers: {
        authorization: 'Bearer ben@example.com-abc123'
      }
    }, function (err, user) {
      authorizer.end()
      session.delete('ben@example.com-abc123')
      session.unlock('ben@example.com-abc123')

      t.equal(err, null)
      t.equal(user.email, 'ben@example.com')
      t.equal(user.cacheAllowed, false)
      t.end()
    })
  })
})

tap.test('after', function (t) {
  session.end(true)
  t.end()
})


test/authenticator.js
var client = require('redis').createClient()
client.unref()

var Authenticator = require('../lib/authenticator')
var tap = require('tap')

var credentials = {
  name: 'foo',
  password: 'bar',
  email: 'ben@example.com'
}

process.env.OAUTH2_AUTHORIZATION_PATH = 'https://auth.example.com'

tap.test('it generates a token and returns a session on login', function (t) {
  var authenticator = new Authenticator()

  authenticator.authenticate({
    body: credentials
  }, function (err, s) {
    authenticator.end()

    t.assert(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/.test(s.token))
    t.equal(s.user.name, 'foo')
    t.equal(s.user.email, 'ben@example.com')
    t.assert(/https:\/\/auth\.example\.com/.test(s.user.sso))
    t.equal(err, null)
    t.end()
  })
})

tap.test('after', function (t) {
  client.end(true)
  t.end()
})


lib/authorizer.js
var _ = require('lodash')
var request = require('request')
var Session = require('./session')

function AuthorizerOAuth2 (opts) {
  _.extend(this, {
    session: new Session(opts),
    // not used in auth plugin.
    profileUrl: process.env.OAUTH2_PROFILE
  }, opts)
}

AuthorizerOAuth2.prototype.authorize = AuthorizerOAuth2.prototype.whoami = function (credentials, cb) {
  var _this = this
  if (!validateCredentials(credentials)) return cb(error404())
  var token = credentials.headers.authorization.replace('Bearer ', '')
  this.session.get(token, function (err, user) {
    if (err) return cb(err)
    else if (!user.accessToken) return _this.session.oauthURL(cb, token)
    else {
      // since this plugin manages its own caching, let the upstream service
      // know that it should not cache results itself
      user.cacheAllowed = false
      // we hold a lock in redis for a few minutes, to prevent
      // a thundering herd of auth requests.
      _this.session.checkLock(token, function (locked) {
        if (locked) {
          return cb(null, user)
        } else {
          _this._checkToken(user.accessToken, function (err) {
            if (err) {
              return _this.session.oauthURL(cb, token)
            } else {
              _this.session.lock(token)
              return cb(null, user)
            }
          })
        }
      })
    }
  })
}

AuthorizerOAuth2.prototype._checkToken = function (accessToken, cb) {
  request.get({
    url: this.profileUrl,
    headers: {
      'user-agent': 'npm Enterprise'
    },
    auth: {
      bearer: accessToken
    },
    json: true
  }, function (err, res, body) {
    if (res && res.statusCode >= 400) err = Error('token no longer valid status = ' + res.statusCode)
    if (err) return cb(err)
    else return cb(null)
  })
}

function validateCredentials (credentials) {
  if (!credentials) return false
  if (!credentials.headers) return false
  if (!credentials.headers.authorization || !credentials.headers.authorization.match(/Bearer /)) return false
  return true
}

function error404 () {
  var error = Error('not found')
  error.statusCode = 404
  return error
}

AuthorizerOAuth2.prototype.end = function () {
  this.session.end()
}

module.exports = AuthorizerOAuth2


lib/authenticator.js
var _ = require('lodash')
var Session = require('./session')
var uuid = require('uuid')

function AuthenticatorOAuth2 (opts) {
  _.extend(this, {
    session: new Session(opts)
  }, opts)
}

AuthenticatorOAuth2.prototype.authenticate = function (credentials, cb) {
  if (!validateCredentials(credentials)) return cb(error500())

  var token = uuid.v4()

  this.session.oauthURL(function (_, url) {
    return cb(null, {
      token: token,
      user: {
        name: credentials.body.name,
        email: credentials.body.email,
        sso: url
      }
    })
  }, token)
}

AuthenticatorOAuth2.prototype.unauthenticate = function (token, cb) {
  this.session.delete(token, cb)
}

function validateCredentials (credentials) {
  if (!credentials) return false
  if (!credentials.body) return false
  if (!credentials.body.name) return false
  return true
}

AuthenticatorOAuth2.prototype.end = function () {
  this.session.end(true)
}

function error500 () {
  var error = Error('unknown error')
  error.statusCode = 500
  return error
}

module.exports = AuthenticatorOAuth2


lib/session.js
var _ = require('lodash')
var oauth2 = require('simple-oauth2')
var redis = require('redis')

function SessionOAuth2 (opts) {
  _.extend(this, {
    sessionLookupPrefix: 'user-',
    sessionLockPrefix: 'user-token-lock-',
    lockTime: toInt(process.env.AUTHZ_CACHE_DEFAULT_TTL_SECONDS, 300),
    client: redis.createClient(process.env.LOGIN_CACHE_REDIS),
    clientId: process.env.OAUTH2_CLIENT_ID,
    clientSecret: process.env.OAUTH2_CLIENT_SECRET,
    site: process.env.OAUTH2_SITE,
    tokenPath: process.env.OAUTH2_TOKEN_PATH,
    authorizationPath: process.env.OAUTH2_AUTHORIZATION_PATH,
    redirectUri: process.env.OAUTH2_REDIRECT_URI,
    scope: process.env.OAUTH2_SCOPE,
    // not used in auth plugin.
    profileUrl: process.env.OAUTH2_PROFILE,
    emailKey: process.env.OAUTH2_EMAIL_KEY,
    userKey: process.env.OAUTH2_USER_KEY
  }, opts)

  this.oauth2 = oauth2({
    clientID: this.clientId,
    clientSecret: this.clientSecret,
    site: this.site,
    tokenPath: this.tokenPath,
    authorizationPath: this.authorizationPath
  })
}

SessionOAuth2.prototype.get = function (key, cb) {
  var _this = this

  key = normalizeKey(key)

  this.client.get(_this.sessionLookupPrefix + key, function (err, user) {
    if (err) return cb(error500())
    if (!user) return _this.oauthURL(cb)

    return cb(null, JSON.parse(user))
  })
}

SessionOAuth2.prototype.checkLock = function (key, cb) {
  var _this = this

  key = normalizeKey(key)

  this.client.get(_this.sessionLockPrefix + key, function (_err, lock) {
    if (lock) return cb(true)
    else return cb(false)
  })
}

SessionOAuth2.prototype.set = function (key, user, cb) {
  var _this = this

  key = normalizeKey(key)

  _this.client.set(this.sessionLookupPrefix + key, JSON.stringify(user), cb)
}

SessionOAuth2.prototype.lock = function (key) {
  var _this = this

  key = normalizeKey(key)

  _this.client.setex(this.sessionLockPrefix + key, this.lockTime, 'locked')
}

SessionOAuth2.prototype.unlock = function (key) {
  var _this = this

  key = normalizeKey(key)

  _this.client.del(this.sessionLockPrefix + key)
}

SessionOAuth2.prototype.delete = function (key, cb) {
  key = normalizeKey(key)

  this.client.del(this.sessionLookupPrefix + key, this.sessionTimeoutPrefix + key, cb)
}

function normalizeKey (token) {
  return token.replace(/^user-/, '')
}

SessionOAuth2.prototype.end = function () {
  this.client.end(true)
}

SessionOAuth2.prototype.oauthURL = function (cb, relayState) {
  // Authorization uri definition
  var authUrl = this.oauth2.authCode.authorizeURL({
    redirect_uri: this.redirectUri,
    scope: this.scope,
    state: relayState
  })

  process.nextTick(function () {
    var error = Error('visit ' + authUrl + ' to validate your session')
    error.statusCode = 401
    return cb(error, authUrl)
  })
}

function error500 () {
  var error = Error('unknown error')
  error.statusCode = 500
  return error
}

function toInt (val, _default) {
  const integer = parseInt(val, 10)
  return isNaN(integer) ? _default : integer
}

module.exports = SessionOAuth2
